File: typescript/packages/core/src/commands/builtin/generic_bind/builders/od.ts
import { odGeneric, parseCount } from '../../generic/od.ts'
import { resolveSource } from '../../utils/stream.ts'
import { specOf } from '../../../spec/builtins.ts'
import { FlagView } from '../../../spec/types.ts'
import { type Builder, resolveGlobOf } from '../adapter.ts'

async function* concatSources(sources: AsyncIterable<Uint8Array>[]): AsyncIterable<Uint8Array> {
  for (const source of sources) {
    for await (const chunk of source) yield chunk
  }
}

export const OD_BUILDER: Builder = {
  name: 'od',
  read: true,
  fn: async (ops, accessor, paths, _texts, opts) => {
    const index = opts.index ?? undefined
    const resolved = paths.length > 0 ? await resolveGlobOf(ops)(accessor, paths, index) : []
    // od defines multiple FILE operands as one concatenated input, so skip
    // and limit offsets apply across the whole run, not per file.
    const source =
      resolved.length === 0
        ? resolveSource(opts.stdin)
        : concatSources(resolved.map((p) => ops.readStream(accessor, p, index)))
    const fl = new FlagView(opts.flags, specOf('od'))
    // asStr, not a truthiness test: an explicitly empty value is an
    // invalid argument to GNU, not an absent flag.
    const skipValue = fl.asStr('skip_bytes')
    const limitValue = fl.asStr('read_bytes')
    return odGeneric(
      source,
      fl.asStr('address_radix') ?? 'o',
      skipValue !== undefined ? parseCount(skipValue, '-j') : 0,
      limitValue !== undefined ? parseCount(limitValue, '-N') : null,
      fl.asList('format'),
    )
  },
}

File: python/mirage/commands/builtin/generic_bind/builders/od.py
from mirage.accessor.base import Accessor
from mirage.commands.builtin.generic.od import od as generic_od
from mirage.commands.builtin.generic.od import parse_count
from mirage.commands.builtin.generic_bind.adapter import (Builder, CommandIO,
                                                          bound_op)
from mirage.commands.builtin.generic_bind.builders.common import \
    resolve_or_empty
from mirage.commands.config import CommandOpts
from mirage.commands.spec import SPECS
from mirage.commands.spec.types import FlagView
from mirage.io.types import ByteSource, IOResult
from mirage.types import PathSpec


async def od(ops: CommandIO, accessor: Accessor, paths: list[PathSpec],
             texts: list[str],
             opts: CommandOpts) -> tuple[ByteSource | None, IOResult]:
    fl = FlagView(opts.flags, spec=SPECS["od"])
    paths = await resolve_or_empty(ops, accessor, paths, opts.index)
    formats = fl.as_list("format")
    # as_str, not `x or y`: the latter would swallow an explicitly empty
    # value, which GNU rejects loudly (`od -N ''` is an invalid argument,
    # not an absent flag).
    skip_value = fl.as_str("skip_bytes")
    limit_value = fl.as_str("read_bytes")
    return await generic_od(
        paths,
        read_stream=bound_op(ops.read_stream, accessor, opts.index),
        stdin=opts.stdin,
        address_radix=fl.as_str("address_radix") or "o",
        skip=(parse_count(skip_value, "-j") if skip_value is not None else 0),
        limit=(parse_count(limit_value, "-N")
               if limit_value is not None else None),
        formats=formats,
    )


BUILDER = Builder("od", od, read=True)
